[GTK] Simplify handling of toplevel window signals

carlosgcampos opened the pull request:
#### c4614fd4f2f2fff205c92dcabd1879feca1b728e
<pre>
[GTK] Simplify handling of toplevel window signals
<a href="https://bugs.webkit.org/show_bug.cgi?id=262944">https://bugs.webkit.org/show_bug.cgi?id=262944</a>

Reviewed by Alejandro G. Castro.

We currently track the state of the toplevel window from every web view,
connecting to the same signals to be notified when something changes. We
could add a toplevel window wrapper instead to track its state and
notify all the web views when there&apos;s any change.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/gtk/ToplevelWindow.cpp: Added.
(WebKit::toplevelWindows):
(WebKit::toplevelWindowDestroyedCallback):
(WebKit::ToplevelWindow::forGtkWindow):
(WebKit::ToplevelWindow::ToplevelWindow):
(WebKit::ToplevelWindow::~ToplevelWindow):
(WebKit::ToplevelWindow::addWebView):
(WebKit::ToplevelWindow::removeWebView):
(WebKit::ToplevelWindow::isActive const):
(WebKit::ToplevelWindow::isFullscreen const):
(WebKit::ToplevelWindow::connectSignals):
(WebKit::ToplevelWindow::disconnectSignals):
(WebKit::ToplevelWindow::connectSurfaceSignals):
(WebKit::ToplevelWindow::disconnectSurfaceSignals):
(WebKit::ToplevelWindow::notifyIsActive):
(WebKit::ToplevelWindow::notifyState):
(WebKit::ToplevelWindow::notifyMonitorChanged):
* Source/WebKit/UIProcess/API/gtk/ToplevelWindow.h: Added.
(WebKit::ToplevelWindow::window const):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseToplevelWindowIsActiveChanged):
(webkitWebViewBaseToplevelWindowStateChanged):
(webkitWebViewBaseToplevelWindowMonitorChanged):
(webkitWebViewBaseSetToplevelOnScreenWindow):
(webkitWebViewBaseDispose):
(webkitWebViewBaseMap):
(webkitWebViewBaseRoot):
(webkitWebViewBaseUnroot):
(webkitWebViewBaseHierarchyChanged):
(webkitWebViewBaseToplevelOnScreenWindowIsFullScreen):
(webkitWebViewBaseEnterFullScreen):
(webkitWebViewBaseExitFullScreen):
(webkitWebViewBaseTakeViewSnapshot):
(toplevelWindowFocusInEvent): Deleted.
(toplevelWindowFocusOutEvent): Deleted.
(toplevelWindowStateEvent): Deleted.
(toplevelWindowConfigureEvent): Deleted.
(toplevelWindowRealized): Deleted.
(toplevelWindowIsActiveChanged): Deleted.
(toplevelWindowStateChanged): Deleted.
(toplevelWindowMonitorsChanged): Deleted.
(toplevelWindowUnrealized): Deleted.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:

Canonical link: <a href="https://commits.webkit.org/269136@main">https://commits.webkit.org/269136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bda25b69e265aa84112de698c90f66d90021398a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21709 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23575 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20099 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/26187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22241 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21937 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/26187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/18807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24429 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/26187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25952 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/26187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/19873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23811 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20352 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19680 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23905 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2687 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alexgcastro reviewed (approved):
LGTM! Great refactoring!

carlosgcampos committed:
[GTK] Simplify handling of toplevel window signals
https://bugs.webkit.org/show_bug.cgi?id=262944

Reviewed by Alejandro G. Castro.

We currently track the state of the toplevel window from every web view,
connecting to the same signals to be notified when something changes. We
could add a toplevel window wrapper instead to track its state and
notify all the web views when there's any change.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/gtk/ToplevelWindow.cpp: Added.
(WebKit::toplevelWindows):
(WebKit::toplevelWindowDestroyedCallback):
(WebKit::ToplevelWindow::forGtkWindow):
(WebKit::ToplevelWindow::ToplevelWindow):
(WebKit::ToplevelWindow::~ToplevelWindow):
(WebKit::ToplevelWindow::addWebView):
(WebKit::ToplevelWindow::removeWebView):
(WebKit::ToplevelWindow::isActive const):
(WebKit::ToplevelWindow::isFullscreen const):
(WebKit::ToplevelWindow::connectSignals):
(WebKit::ToplevelWindow::disconnectSignals):
(WebKit::ToplevelWindow::connectSurfaceSignals):
(WebKit::ToplevelWindow::disconnectSurfaceSignals):
(WebKit::ToplevelWindow::notifyIsActive):
(WebKit::ToplevelWindow::notifyState):
(WebKit::ToplevelWindow::notifyMonitorChanged):
* Source/WebKit/UIProcess/API/gtk/ToplevelWindow.h: Added.
(WebKit::ToplevelWindow::window const):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseToplevelWindowIsActiveChanged):
(webkitWebViewBaseToplevelWindowStateChanged):
(webkitWebViewBaseToplevelWindowMonitorChanged):
(webkitWebViewBaseSetToplevelOnScreenWindow):
(webkitWebViewBaseDispose):
(webkitWebViewBaseMap):
(webkitWebViewBaseRoot):
(webkitWebViewBaseUnroot):
(webkitWebViewBaseHierarchyChanged):
(webkitWebViewBaseToplevelOnScreenWindowIsFullScreen):
(webkitWebViewBaseEnterFullScreen):
(webkitWebViewBaseExitFullScreen):
(webkitWebViewBaseTakeViewSnapshot):
(toplevelWindowFocusInEvent): Deleted.
(toplevelWindowFocusOutEvent): Deleted.
(toplevelWindowStateEvent): Deleted.
(toplevelWindowConfigureEvent): Deleted.
(toplevelWindowRealized): Deleted.
(toplevelWindowIsActiveChanged): Deleted.
(toplevelWindowStateChanged): Deleted.
(toplevelWindowMonitorsChanged): Deleted.
(toplevelWindowUnrealized): Deleted.
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:

Canonical link: https://commits.webkit.org/269136@main
